[Spring 체스 - 2단계] 애쉬(김동호) 미션 제출합니다.

dongho108 opened the pull request:
안녕하세요 제이! 2단계 미션 제출합니다.
DM으로 질문 많이 드렸는데 답장 많이 해주셔서 감사합니다 ㅠㅠ!!

### 1단계 피드백 반영

- ChessGame이 필드에 존재
> 여러 게임을 동시에 실행시킬 수 없으므로 id를 이용해 매번 가져오도록 수정했습니다.

- repository vs dao
> 지금까지 둘의 차이를 아예 모르고 용어를 사용하고 있었습니다.
그래서 일단 느낌이라도 알려고 간단하게 학습을 했습니다.
지금은 repository는 특정 엔티티 중심, dao는 db중심이라는 정도로만 알고있고, 이번 미션 동안 더 학습해 나갈 예정입니다.


### 2단계 구현

> UI는 정말 부족하게 짰습니다. 얼른 피드백받고 더 공부하려고 최대한 빨리 작성해 제출합니다!

dongho108 committed:
chore: mysql 최초 디비 세팅 수정

dongho108 committed:
docs: 2단계 기능목록 추가

dongho108 committed:
refactor: chess game을 game_name이 아닌 id로 조회할 수 있도록 변경

- 테이블구조 변경
- url 매핑 변경
- dao query 변경
- ChessService 에서 ChessGame을 더이상 필드로 들고있지 않도록 변경

dongho108 committed:
refactor: 재시작 기능 제거

- 로비로 가기 기능으로 대체 가능하기 때문

dongho108 committed:
refactor: chessGameDao 테스트 추가

- 메모리db h2 기반으로 테스트
- @SpringBootTest, @Transactional을 통해 트랜잭션 테스트

dongho108 committed:
style: 주석 제거

dongho108 committed:
fix: 게임종료가 DB에 저장되지 않던 문제 해결

dongho108 committed:
test: ChessService 테스트 추가

dongho108 committed:
feat: 방 생성시 입력했던 비밀번호를 입력해야 삭제할 수 있는 기능 추가

dongho108 committed:
docs: 구현된 요구사항 표시

dongho108 committed:
refactor: 반복되는 html 코드를 하나로 줄임

- piece dto 에 position 를 추가하여 보내줌
- model에 piece dto list를 담아 보내서 템플릿 엔진으로 반복하여 처리

dongho108 committed:
remove: 필요없는 클래스 제거

dongho108 committed:
refactor: 필요없는 메서드 제거

- 기존의 html에서 체스를 보여주던 방식이 바뀌어 getCurrentChessBoard가 필요가 없어짐

dongho108 committed:
refactor: 반환값을 사용하지 않는 메서드의 반환타입 void로 변경

dongho108 committed:
style: 주석 제거

dongho108 committed:
test: ChessService delete 테스트 추가

jjj0611 commented on src/main/java/chess/dao/ChessGameDao.java:
```diff
@@ -1,53 +1,82 @@
 public class ChessGameDao {
 
-    private JdbcTemplate jdbcTemplate;
+    private final JdbcTemplate jdbcTemplate;
```
한 번 주입한 뒤 변경할 수 없도록 잘 막아주셨네요 👍

jjj0611 commented on src/main/java/chess/dao/ChessGameDao.java:
```diff
@@ -1,53 +1,82 @@
+            PreparedStatement ps = connection.prepareStatement(sql, new String[] {"id"});
+            ps.setString(1, gameName);
+            ps.setString(2, password);
+            ps.setString(3, "white");
```
항상 첫 턴이 화이트여서 이렇게 저장하는걸까요?
만약 규칙이 바뀌어서 흑이 먼저 시작하는 걸로 바뀌면 이전에 생성된 게임들은 백이 먼저 시작되고,
새롭게 생성된 게임들은 흑이 먼저 시작되는 구조겠네요 :)

dongho108 replied:
게임이 시작될때 ready상태로 시작되고 게임시작 버튼을 누르면 그때 턴이 저장되도록 수정했습니다. 
만약 첫턴을 블랙으로 바꾸게되면 도메인로직이 수정되는 것이므로 도메인의 execute에서 command가 start시 new Black()을 return하도록 하면 됩니다.

jjj0611 replied:
게임을 처음 생성하고 생성된 도메인에서 턴을 뽑아서 저장하면 안되나요?
위와 같은 방식을 채택한 이유가 궁금합니다.

jjj0611 commented on src/main/java/chess/dao/ChessGameDao.java:
```diff
@@ -1,53 +1,82 @@
+        return keyHolder.getKey().intValue();
+    }
+
+    public List<ChessGameDto> findAllChessGames() {
+        String sql = "select id, game_name, turn from chess_game";
+        return jdbcTemplate.query(sql, chessGameDtoMapper());
+    }
```
체스 게임이 2조 개 생성되어 있으면 어떻게 될까요..?!

dongho108 replied:
너무 많은 데이터를 가져올 것 같습니다.
select 쿼리에 limit 를 두고 가져오도록 수정했습니다.

추가로 100개만 가져오지만 101번쨰 만든 방을 들어가기 위해선 방들을 page로 구분해서 가져오는 기능이 있어야 할 것 같아요.

jjj0611 replied:
네 page에 대해 고민해주셨네요 :)

모든 데이터를 한 번에 불러오는 것은 언제나 위험할 수 있습니다 :)

jjj0611 commented on src/main/java/chess/service/ChessService.java:
```diff
@@ -9,89 +9,85 @@
-    public List<String> move(String from, String to, String gameName) {
-        Command command = Command.from(makeCommand(from, to));
+    public void delete(String password, int chessGameId) {
+        if (isEnd(chessGameId)) {
```
체스 게임이 완전히 끝나지 않으면 삭제를 할 수가 없군요!

dongho108 replied:
요구사항이 종료된 게임만 삭제할 수 있다여서요!!

dongho108 replied:
그래도 게임이 시작되지 않았으면 삭제가 가능하도록 수정했습니다!

jjj0611 commented on src/main/java/chess/dao/ChessGameDao.java:
```diff
@@ -1,53 +1,82 @@
         }, chessGameRowMapper);
 
-        if(result.isEmpty()) {
+        if (result.isEmpty()) {
```
메서드를 호출하는 곳에서 NPE가 터질 가능성은 없나요??

dongho108 replied:
기존의 beforeFirst()를 이용하기위해 preparedStatement로 옵션을 지정해주던 코드를 없애고 단건조회 쿼리를 하도록 수정했습니다.

jjj0611 commented on src/main/java/chess/dao/ChessGameDao.java:
```diff
@@ -94,8 +123,8 @@ private Piece makePiece(ResultSet resultSet) throws SQLException {
-    public void remove(String gameName) {
-        String sql = "delete from chessgame where game_name = ?";
-        jdbcTemplate.update(sql, gameName);
+    public void delete(String password, int chessGameId) {
+        String sql = "delete from chess_game where password = ? and id = ?";
+        jdbcTemplate.update(sql, password, chessGameId);
```
만약 이 과정에서 DataAccessException이 발생하면 어떻게 되나요??

dongho108 replied:
```DataAccessException``` 을 언급하신게

1. 쿼리를 실행하다가 오류가 생기는 것을 의미 하신것인가요?
2. 아니면 정말 ```DataAccessException``` 자체를 언급하신 건가요? ```DataAccessException```을 알아보니 jdbc 템플릿이 그냥 sql exception을 한번 포장해서 전해주는 예외라고 해서요!

만약 1번을 언급하신 거라면 런타임 exception이니 예외페이지로 갈 것 같습니다. 방 삭제하다가 예외페이지로 가는 것은 이상하다고 생각되지만 password가 틀릴경우 예외가 발생하진 않고 아무변화도 일어나지 않는 것을 보니 ```DataAccessException``` 이 터지는 것은 정말 DB연결에 심각한 문제가 터진 것이니 예외페이지로 가는게 맞다고 생각이 됩니다.

제가 질문의 의도를 잘못 파악한 걸까요?!

woowahan-neo replied:
제가 기존에 남겼던 코멘트는 아니지만 제 의견도 남겨둘게요~
패스워드가 틀린 상황인지, 존재하지 않는 게임인지 아니면 DB에 문제가 있는 상황인지 등에 대해 별도로 예외처리 하지 않아도 될지 고민해보세요 :)

jjj0611 replied:
넵 네오의 말에 대한 이야기였습니다.
password와 id가 둘 다 같은 row를 삭제하는 식으로 비밀번호 일치 여부를 확인하게 되면,
DataAccessException이 터졌을 때 패스워드가 일치하지 않은 것인지, 예외가 발생한 상황인지 등 각 상황에 대해 아는 방법이 예외로그를 보고 판단해야하는 문제가 있다고 생각합니다.
어떤 것이 문제가 되는지 케이스를 잘 나누는 것이 중요하다고 생각합니다

jjj0611 commented on src/main/java/chess/service/ChessService.java:
```diff
@@ -9,89 +9,85 @@
+        Command command = Command.from(makeCommand(from, to));
         chessGame.progress(command);
-        pieceDao.deleteByPosition(to, gameName);
-        pieceDao.updatePosition(from, to, gameName);
-        chessGameDao.update(ChessGameDto.from(chessGame));
-        return chessGame.getChessBoardSymbol();
+        pieceDao.deleteByPosition(to, chessGameId);
+        pieceDao.updatePosition(from, to, chessGameId);
+        chessGameDao.update(chessGame.getState().getTurn(), chessGameId);
```
체스 게임을 찾아서 커맨드로 진행시키는 것 까지는 좋은데,
pieceDao에서 삭제하고, 저장하는 로직과 chessGame을 저장하는 로직까지는 개념적으로 하나로 합쳐져 있어야 할 것 같아요.
제가 저번에 말씀드린 58번 라인에서 성공하고, 59번에서 실패하면 어떻게 될까요?와는 별개로 드리는 코멘트에요!

dongho108 replied:
삭제하고 update, update하는 쿼리는 결국에 세번 나가야할 것 같아서  ```개념적으로 합치라```는 말씀에 집중해 그냥 메서드로 묶었습니다.

jjj0611 replied:
엇 단순히 메서드를 묶는 것을 요청드린 것은 아닙니다.
게임이 진행된 뒤에 진행된 게임을 기반으로 db를 갱신하는 것이 아니라
게임이 진행이 잘 되었으니 dao를 요청 받음 값으로 갱신하는 방식에 대해 이야기 드린거에요.
만
어떻게 생각하시나요??

jjj0611 commented on src/main/java/chess/service/ChessService.java:
```diff
@@ -9,89 +9,85 @@
-        chessGame.progress(Command.from("start"));
-
+    public List<String> findChessBoardById(int chessGameId) throws IllegalStateException {
+        ChessGame chessGame = chessGameDao.findById(chessGameId);
```
비밀번호는 게임 참가에 쓰이는게 아니라 삭제에만 쓰이나요???

dongho108 replied:
요구사항이 게임참가일땐 방제목클릭, 삭제할때는 패스워드가 필요하다고 되어있어서 이렇게 구현했어요!

jjj0611 commented on src/test/java/chess/dao/ChessGameDaoTest.java:
```diff
@@ -0,0 +1,90 @@
+import chess.dto.ChessGameDto;
+
+@SpringBootTest
+@Transactional
```
어떤 의미로 붙여주신걸까요??
Transaction에 대한 공부를 하셨을까요??

dongho108 replied:
테스트에서 @Transactional이 데이터를 자동으로 rollback 시키는 기능을 사용하려고 붙였습니다!!

jjj0611 replied:
`@Transactional` 어노테이션을 사용하기 전에 트랜잭션에 대한 개념을 더 먼저 공부하시면 좋을 것 같아요 :)

jjj0611 commented on src/test/java/chess/service/ChessServiceTest.java:
```diff
@@ -0,0 +1,89 @@
+
+@SpringBootTest
+@Transactional
+class ChessServiceTest {
```
전체적으로 Service 테스트가 조금 부실한 것 같아요.
테스트 하지 않은 부분들은 필요가 없다고 느껴서일까요??

dongho108 replied:
단순 도메인로직을 호출하는 메서드는 테스트 하지 않았습니다. 그 외에 추가로 행위를 해주는 메서드들은 테스트를 추가해주었습니다!

jjj0611 reviewed (changes requested):
안녕하세요 애쉬! 리뷰어제이입니다.

구현해주신 코드 잘 봤습니다!
간단하게 고민해볼만한 부분들 코멘트로 남겨두었어요.

코멘트 확인부탁드리고, 혹시 궁금하신 점이나 잘 모르겠다 싶은 부분은 DM이나 코멘트로 알려주세요 :)

dongho108 committed:
refactor: findAll 쿼리에 limit 추가

dongho108 committed:
refactor: move관련 DB저장 로직 하나의 메서드로 합침

dongho108 committed:
test: @DisplayName 추가 및 findById 테스트 추가

dongho108 committed:
test: finish, move 테스트 추가

dongho108 committed:
refactor: 조회해올때 queryForObject를 사용하도록 수정

- beforeFirst()를 이용하기위해 preparedStatement로 옵션을 지정해주던 코드를 없앰

dongho108 committed:
refactor: 게임 Start 버튼/기능 추가

- 게임생성시 ready 상태로 저장되고 start시 턴이 결정되도록 수정

dongho108 committed:
refactor: 어색한 논리표현 수정

dongho108 committed:
refactor: 게임이 ready 상태일 경우에도 삭제가 가능하도록 수정

woowahan-neo commented on README.md:
```diff
@@ -55,17 +55,31 @@
+* [o] localhost:8080 에서 체스방 목록을 조회할 수 있어야 한다.
+  * [o] 체스방 제목과 체스방 삭제버튼이 표시된다.
+* [o] 체스방 목록에서 체스방 제목을 클릭하면 체스 게임을 이어서 진행할 수 있어야 한다.
+* [ ] 체스방 삭제버튼 클릭시 비밀번호를 입력해 체스게임을 삭제한다.
+  * [ ] 진행중인 체스방은 삭제할 수 없다.
```
```suggestion
* [o] 체스방 삭제버튼 클릭시 비밀번호를 입력해 체스게임을 삭제한다.
  * [o] 진행중인 체스방은 삭제할 수 없다.
```

체크를 놓치셨네요~

woowahan-neo commented on build.gradle:
```diff
@@ -24,6 +24,7 @@ dependencies {
     testImplementation 'io.rest-assured:rest-assured:4.4.0'
 
     runtimeOnly 'mysql:mysql-connector-java:8.0.28'
+	runtimeOnly 'com.h2database:h2'
```
```suggestion
    runtimeOnly 'com.h2database:h2'
```

매의 눈 @_@

dongho108 replied:
저것만 tab이어서 인텔리제이에서 안보였네요..;;; ㅎㅎ

woowahan-neo commented on src/test/java/chess/service/ChessServiceTest.java:
```diff
@@ -0,0 +1,103 @@
+        String teamName = chessService.finish(Command.from("end"), savedId);
+        assertThat(teamName).isEqualTo("DRAW");
+    }
+}
```
```suggestion
}

```

EOF를 챙겨주세요~

- 여러 버전의 유닉스 OS간의 공통된 API 및 인터페이스를 정의해놓은 POSIX 명세 중 text file 부분을 보면 유닉스 시스템은 text file들을 서로 구분할때 EOL을 사용하기 때문에 파일을 만들때마다 파일의 끝에 EOL을 넣어주는 것이 안전하다.

dongho108 replied:
이거 딴 사람꺼는 잘보이는데 제껀 못봤네요..ㅋㅋ

woowahan-neo reviewed (changes requested):
미션 잘 진행해주셨습니다 👍 
소소한 코멘트 남겨두었으니 확인 부탁드릴게요~

dongho108 committed:
docs: 기능 완료 목록 체크

dongho108 committed:
style: indent tab 대신 4spaces 사용하도록 수정

dongho108 committed:
style: 파일 끝에 EOL 넣어줌

jjj0611 commented on src/main/java/chess/dao/ChessGameDao.java:
```diff
@@ -1,83 +1,90 @@
+            PreparedStatement ps = connection.prepareStatement(sql, new String[] {"id"});
+            ps.setString(1, gameName);
+            ps.setString(2, password);
+            ps.setString(3, "ready");
```
게임은 항상 ready 상태로 save되겠네요

dongho108 replied:
전달받은 엔티티의 값으로 세팅해주도록 변경했습니다!!

jjj0611 commented on src/main/java/chess/dao/ChessGameDao.java:
```diff
@@ -1,83 +1,90 @@
-    public void update(ChessGameDto chessGameDto) {
-        String sql = "update chessgame set turn = ? where game_name = ?";
-        jdbcTemplate.update(sql, chessGameDto.getTurn(), chessGameDto.getGameName());
+    public List<ChessGameDto> findAllChessGames() {
+        String sql = "select id, game_name, turn from chess_game limit 100";
+        return jdbcTemplate.query(sql, chessGameDtoMapper());
```
dao에서 data를 가져올 때 dto를 return 하는 이유는 뭘까요??

dongho108 replied:
처음엔 도메인에 id가 포함되어 있지 않아서 id가 포함된 ChessGameDto를 직접 반환해주도록 했었습니다.
그리고 어차피 컨트롤러에 DTO를 반환해줄거여서 DTO를 반환했었습니다.

하지만 DAO는 컨트롤러의 존재도 모르고 DTO의 존재도 모르기 때문에 DTO를 반환해주면 안됩니다.
그래서 List<ChessGame>을 반환하도록 수정했습니다.

jjj0611 reviewed (changes requested):
안녕하세요 애쉬! 리뷰어제이입니다.

구현해주신 코드 잘 봤습니다!
코멘트 조금 남겨드렸으니 확인 부탁드려요!

이해가지 않는 부분은 코멘트나 DM으로 해주셔도 됩니다 :)

dongho108 committed:
refactor: 게임삭제시 발생하는 예외 세분화

dongho108 committed:
refactor: save 기능을 dao가 domain 데이터를 기반으로 업데이트 하도록 수정

dongho108 committed:
refactor: 각 계층에 맞게 반환타입 수정

- ChessGameDao의 findAllChessGames()에서 List<ChessGame>을 반환하도록 수정
- ChessService에서는 List<ChessGame>을 List<ChessGameDto>로 변환해 반환하도록 수정

dongho108 committed:
test: Transactional 어노테이션 제거

- BeforeEach로 매번 데이터를 지우고 새로 세팅하도록 수정

dongho108 committed:
refactor: IllegalArgumentException을 한 곳에서 처리하도록 수정

jjj0611 commented on src/main/java/chess/service/ChessService.java:
```diff
@@ -2,96 +2,113 @@
-        chessGameDao.update(ChessGameDto.from(chessGame));
-        return chessGame.getChessBoardSymbol();
-    }
+        if (chessGame.isProgress()) {
```
한 메서드의 라인 수가 10줄을 넘어가지 않는다 는 규칙은 더 이상 유효하지 않은 걸까요??

jjj0611 commented on src/main/java/chess/service/ChessService.java:
```diff
@@ -2,96 +2,113 @@
         return chessGame.isEnd();
     }
+
+    public List<PieceDto> getPieces(int chessGameId) {
```
getPieces와 findChessGame에서
어떤 경우에는 get을 어떤 경우에는 find를 쓴걸까요??

jjj0611 commented on src/main/java/chess/controller/ControllerExceptionHandler.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    @ExceptionHandler(IllegalArgumentException.class)
+    public String handleIllegalArgumentException(RedirectAttributes redirectAttributes,
```
이 핸들러에서만 error log를 찍지 않는 것 같은데
애쉬만의 규칙이 있는걸까요?

jjj0611 commented on src/main/java/chess/controller/SpringController.java:
```diff
@@ -12,9 +12,10 @@
+import chess.dto.PieceDto;
 import chess.service.ChessService;
 
 @Controller
```
controller의 이름이 조금 아쉽네요.
Spring을 도입했기 때문에 스프링 컨트롤러일까요?

jjj0611 commented on src/main/java/chess/domain/ChessGame.java:
```diff
@@ -90,6 +104,10 @@ private double calculateScore(Team team) {
+        return this.password.equals(password);
+    }
+
     public List<String> getSymbols() {
```
이 부분이 도메인 로직이 맞을까요??

jjj0611 commented on src/main/java/chess/dto/ChessGameDto.java:
```diff
@@ -25,7 +26,11 @@ public static ChessGameDto from(ChessGame chessGame) {
+        return new ChessGameDto(chessGame.getId(), chessBoardDto, gameName, turn);
+    }
+
+    public int getChessGameId() {
```
아래에는 더이상 사용하지 않는 메서드도 많이 있네요!
그대로 남겨두신 이유가 있을까요??
이 객체 말고도 사용하지 않는 메서드가 꽤 있어보입니다!

jjj0611 reviewed (approved):
안녕하세요 애쉬! 리뷰어 제이입니다!

제출해주신 코드 잘 봤습니다 :)
간단하게만 코멘트 몇 개 더 남겼어요!

다른 것들은 보고 넘어가주셔도 되는 것들인데,
도메인에 도메인 로직이 아닌 것 같은 부분들이 좀 보이네요.

그런 부분만 신경 써서 다음 미션 진행할 때도 같이 녹여서 고민해주세요!

코드는 여기서 머지할게요!
고생하셨습니다!
